src/care/host_device_ptr.h

```diff
@@ -196,6 +196,22 @@ namespace care {
          return *reinterpret_cast<host_device_ptr<const T> const *> (this);
       }
 
+#if defined(CARE_ENABLE_BOUNDS_CHECKING)
+      template <class Index>
+      inline void boundsCheck(const Index i) const {
+         if (i < 0 || i >= (Index) MA::size()) {
```

Comment:
what happens if the array is created as one type by accessed as another (allocated a char but accessed as an int)?

Reply:
Good point. Now using the size in bytes divided by the size of the type that it is when this operator is called.